src/worleynoise.rs

```diff
@@ -0,0 +1,205 @@
+            region_map : Vec::new(),
+            cell_map : Vec::new(),
+            cell_map_width : cell_map_width,
+            cell_map_height : cell_map_height,
+        };
+
+        map.init_regions();
+
+        for x in 0..cell_map_width{
+            map.cell_map.push(Vec::new());
+            for y in 0..cell_map_height{
+                let cell = Cell::new(x,y,map.region_map[(x / region_size) as usize][(y / region_size) as usize].clone());
+                        map.cell_map[(x) as usize].push(cell);
+            }
+        }
+        map
+    }
+    fn init_regions(&mut self){
+        let mut rng = rand::thread_rng();
+
+        let regions_x = self.cell_map_width / self.region_size;
+        let regions_y = self.cell_map_height / self.region_size;
+
+        for i in 0..regions_x {
+            self.region_map.push(Vec::new());
+            for j in 0..regions_y {
+                let mut region =Region::new(i,j);
+                let xcord = rng.gen_range(0..self.region_size);
+                let ycord = rng.gen_range(0..self.region_size);
+                let node = Node::new(xcord + i*self.region_size ,ycord + j*self.region_size );
+                region.node = Some(node);
+
+                let  rcregion = Rc::new(region);
+
+                self.region_map[i as usize].push(rcregion);
+            }
+        }
+    }
+
+    fn get_regions_in_bound(&self, x : i32, y : i32, radius : i32) -> Vec<&Region>{
+        let mut regions = Vec::new();
+        let x_min = x - radius;
+        let x_max = x + radius;
+        let y_min = y - radius;
+        let y_max = y + radius;
+        for i in x_min..x_max {
+            for j in y_min..y_max {
+
+                let region_x = i;
+                let region_y = j;
+                if region_x != x && region_y != y {
+
+
+                    if region_x >= 0 && region_x < self.region_map.len() as i32 && region_y >= 0 && region_y < self.region_map[region_x as usize].len() as i32 {
+                        let region = &self.region_map[region_x as usize][region_y as usize];
+                        regions.push(region.as_ref());
+                    }
+                }else{
+                    continue;
+                }
+            }
+        }
+        regions
+    }
+
+    fn generate_noise(&mut self,threshold : f32){
+        for i in 0..self.cell_map.len() {
+            for j in 0..self.cell_map[i as usize].len() {
+                let cell =  &self.cell_map[i as usize][j as usize];
+                let region = &self.region_map[cell.region.as_ref().reg_x as usize][cell.region.as_ref().reg_y as usize];
+                let neighbours = self.get_regions_in_bound(region.reg_x,region.reg_y,3);
+
+
+                let mut node_vec = Vec::new();
+                node_vec.push(region.as_ref().node.as_ref().unwrap());
+                for neighbour in neighbours {
+                    let node = neighbour.node.as_ref().unwrap();
+                    node_vec.push(node);
```

Comment:
Can this end up inserting duplicate references into `node_vec`? Would that even be a problem?

Overall it looks like a better approach. I can look properly tomorrow.

Reply:
fixed